src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -753,21 +764,49 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
             {
                 var index = command.Index;
                 var options = mergedCommands.Options[index];
-                Task indexTask = null;
+                var taskResult = new ClusterTransactionResult
+                {
+                    GeneratedResult = mergedCommands.Replies[index],
+                };
                 if (options.WaitForIndexesTimeout != null)
                 {
-                    indexTask = BatchHandlerProcessorForBulkDocs.WaitForIndexesAsync(this, options.WaitForIndexesTimeout.Value,
+                    var indexTask = BatchHandlerProcessorForBulkDocs.WaitForIndexesAsync(this, options.WaitForIndexesTimeout.Value,
                         options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
                             mergedCommands.LastDocumentEtag, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
-                }
 
-                var result = new ClusterTransactionCompletionResult
+                    var removeTask = ServerStore.Cluster.ClusterTransactionWaiter.CreateTask(command.Options.TaskId);
+
+                    /*
+                        the remove task that we are creating here is relevant only for a failover with wait for index:
+                        We need the task only when we do a failover during cluster transaction (batch handler) with 'wait for index'.
+                        Only then we will want the task that created here, because we'll want to 'wait for index' in the new node after the failover.
+                    */
+
+                    indexTask.ContinueWith(t =>
+                    {
+                        if(t.IsFaulted)
+                        {
+                            Exception e = t.Exception.InnerException;
+                            RachisLogIndexNotifications.NotifyListenersAbout(index, e);
+                            ServerStore.Cluster.ClusterTransactionWaiter.SetException(options.TaskId, e);
+                        }
+                        else
+                        {
+                            RachisLogIndexNotifications.NotifyListenersAbout(index, null);
```

Comment:
This is wrong, assume that there is _another_ notification about a *later* index while we are waiting for the indexes to complete, what happens then?

Reply:
Done, I moved it out from the 'if else' scopes